### PR TITLE
fix: Accept non-negative int literals and py expressions as nats

### LINKED ISSUE #704: Non-negative literals should check against `nat`

### PR DESCRIPTION
**Pull request by mark-koch**
Fixes #704

### CODE REVIEW

File: guppylang/std/builtins.py [hunk 1]
```diff
@@ -106,7 +106,9 @@ def __int__(self: bool) -> int:
     @guppy
     @no_type_check
     def __nat__(self: bool) -> nat:
-        # TODO: Literals should check against `nat`
+        # TODO: Type information doesn't flow through the `if` expression, so we
+        #  have to insert the `nat` coercions by hand.
+        #  See https://github.com/CQCL/guppylang/issues/707
         return nat(1) if self else nat(0)
```
Comment (mark-koch) [inline guppylang/std/builtins.py:112]: Sadly, this still fails due to #707

File: guppylang/checker/expr_checker.py [hunk 2]
```diff
@@ -1138,25 +1151,43 @@ def eval_py_expr(node: PyExpr, ctx: Context) -> Any:
...
+        # Only resolve `int` to `nat` if the user specifically asked for it
+        case int(n) if type_hint == nat_type() and n >= 0:
+            return nat_type()
+        # Otherwise, default to `int` for consistency with Python
         case int():
-            return cast(TypeDef, globals["int"]).check_instantiate([], globals)
+            return int_type()
         case float():
-            return cast(TypeDef, globals["float"]).check_instantiate([], globals)
+            return float_type()
```
Comment (mark-koch) [inline guppylang/checker/expr_checker.py:1175]: Drive-by: No need to look these up in globals

File: guppylang/checker/expr_checker.py [hunk 3]
```diff
@@ -1138,25 +1151,43 @@ def eval_py_expr(node: PyExpr, ctx: Context) -> Any:
     return python_val
 
 
-def python_value_to_guppy_type(v: Any, node: ast.expr, globals: Globals) -> Type | None:
+def python_value_to_guppy_type(
+    v: Any, node: ast.expr, globals: Globals, type_hint: Type | None = None
+) -> Type | None:
     """Turns a primitive Python value into a Guppy type.
 
+    Accepts an optional `type_hint` for the expected expression type that is used to
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/checker/expr_checker.py:1159]: Currently it's not an internal error, the hint is just ignored if it's invalid. The actual checking that the synthesised type matches what's expected happens in the `ExprChecker`.

A wider refactor of this would be good, but imo too much for this PR.

**Review by ss2165 — APPROVED** [review]
